src/adapters/wideorbit.js

```diff
@@ -0,0 +1,213 @@
+    var commonParams = _setCommonParams(pos, params);
+
+    if (params.pId) {
+      return _setupIdPlacementParameters(pos, params) + commonParams;
+    }
+
+    return _setupNamedPlacementParameters(pos, params) + commonParams;
+  }
+
+  function _callBids(params) {
+    var publisherId,
+      bidUrl = '',
+      i;
+
+    bids = params.bids || [];
+
+    for (i = 0; i < bids.length; i++) {
+      var requestParams = bids[i].params;
+
+      requestParams.tagId = bids[i].placementCode;
+
+      _fixParamNames(requestParams);
+
+      publisherId = requestParams.pbId;
+      bidUrl += _setupPlacementParameters(i, requestParams);
+    }
+
+    bidUrl = _setupAdCall(publisherId, bids.length, bidUrl);
+
+    utils.logMessage('Calling WO: ' + bidUrl);
+
+    adloader.loadScript(bidUrl);
+  }
+
+  function _processUserMatchings(userMatchings) {
+    var headElem = document.getElementsByTagName('head')[0],
+        createdElem;
+
+    utils._each(userMatchings, function (userMatching) {
+      switch (userMatching.Type) {
+        case 'redirect':
+          createdElem = document.createElement('img');
+          break;
+        case 'iframe':
+          createdElem = utils.createInvisibleIframe();
+          break;
+        case 'javascript':
+          createdElem = document.createElement('script');
+          createdElem.type = 'text/javascript';
+          createdElem.async = true;
+          break;
+      }
+      createdElem.src = decodeURIComponent(userMatching.Url);
+      headElem.insertBefore(createdElem, headElem.firstChild);
+    });
+  }
+
+  function _getBidResponse(id, placements) {
+    var i;
+
+    for (i = 0; i < placements.length; i++) {
+      if (placements[i].ExtPlacementId === id) {
+        return placements[i];
+      }
+    }
+  }
+
+  function _isUrl(scr) {
+    return scr.slice(0, 6) === "http:/" || scr.slice(0, 7) === "https:/" || scr.slice(0, 2) === "//";
+  }
+
+  function _buildAdCode(placement) {
+    var adCode = placement.Source, pixelTag;
+
+    utils._each(placement.TrackingCodes, function (trackingCode) {
+      if (_isUrl(trackingCode)) {
+        pixelTag = '<img src="' + trackingCode + '" width="0" height="0" style="position:absolute"></img>';
+      } else {
+        pixelTag = trackingCode;
+      }
+      adCode = pixelTag + adCode;
+    });
+
+    return adCode;
+  }
+
+  window.pbjs = window.pbjs || {};
+  window.pbjs.handleWideOrbitCallback = function (response) {
+    var bidResponse,
+      bidObject;
+
+    utils.logMessage('WO response. Placements: ' + response.Placements.length);
+
+    _processUserMatchings(response.UserMatchings);
+
+    utils._each(bids, function (bid) {
+      bidResponse = _getBidResponse(bid.placementCode, response.Placements);
+
+      if (bidResponse && bidResponse.Type === 'DirectHTML') {
+        bidObject = bidfactory.createBid(1);
+        bidObject.cpm = bidResponse.Bid;
+        bidObject.ad = _buildAdCode(bidResponse);

```

Comment:
the `bid.ad` should contain the HTML/script markup of the ad. When I tested this, I got empty string for this param, which is invalid. It looks like you might need to get the `MaterialUrl` or `source` param here. If `source` is empty, it should not be a valid bid. Use `bidfactory.createBid(constants.STATUS.NO_BID)` instead.